### PR TITLE
NetworkPolicy unit tests

### PR DESCRIPTION
**Pull request by caseydavenport**
- [x] Validation tests
- [x] storage tests
- [x] strategy tests
- [ ] test-cmd.sh

CC @thockin

### CODE REVIEW

**Comment by caseydavenport** [comment]
Hmm.. currently hitting an error in `hack/test-cmd.sh` on GCE but not locally.

```
+++ [0603 15:17:56] Testing kubectl(v1:networkpolicies)
error validating "docs/user-guide/networkpolicy/networkpolicy.yaml": error validating data: couldn't find type: v1beta1.NetworkPolicy; if you choose to ignore these errors, turn validation off with --validate=false
!!! Error in ./hack/test-cmd.sh:1867
  'kubectl create -f docs/user-guide/networkpolicy/networkpolicy.yaml "${kube_flags[@]}"' exited with status 1
Call stack:
  1: ./hack/test-cmd.sh:1867 runTests(...)
  2: ./hack/test-cmd.sh:2225 main(...)
Exiting with status 1
```

**Comment by caseydavenport** [comment]
Actually, I think I'm getting the above error only when using the binaries run from `_output/local/bin` but not the binaries built in `_output/dockerized/bin`.  I'm not familiar with the differences here - @thockin any thoughts as to what might be going on?

File: pkg/apis/extensions/validation/validation.go [hunk 1]
```diff
@@ -699,7 +699,6 @@ func ValidateNetworkPolicySpec(spec *extensions.NetworkPolicySpec, fldPath *fiel
 		// TODO: Update From to be a pointer to slice as soon as auto-generation supports it.
 		for _, f := range i.From {
 			numFroms := 0
-			allErrs := field.ErrorList{}
```
Comment (caseydavenport) [inline pkg/apis/extensions/validation/validation.go]: This fixes a bug where errors during validation weren't being reported.

If we don't think that the rest of this PR can get in to v1.3, then I think at least the diffs in this file should.

**Comment by caseydavenport** [comment]
@matchstick @girishkalele I hear @thockin has handed off to you while he's on vacation.  Do you think you could take a look at this PR?  It has UTs / stabilization that I think we'd like to have merged for v1.3.

Hoping it should be pretty uncontentious, but still need to figure out what's going on with the jenkins failure - any advice would be much appreciated :)

**Comment by girishkalele** [comment]
@caseydavenport 

Could you fill me in on the background here - am not the one to make the call for 1.3, @matchstick or @thockin would be ones.

If there is a fix for UT errors/flakes in validation.go, maybe we can submit and merge that separately.

pkg/apis/extensions/validation/validation.go

**Comment by caseydavenport** [comment]
@girishkalele Surely.

The background here is that a new NetworkPolicy API resource was merged in https://github.com/kubernetes/kubernetes/pull/25638 and will be new in v1.3

As part of that we agreed that the resource can be merged without UTs with the understanding that I'd put up a tests PR in the near future - this is that PR :)

While writing the tests, I spotted a pair of validation bugs that are currently part of this PR.  I will raise another PR with just the validation fixes as those are more important for the v1.3 milestone.

**Comment by caseydavenport** [comment]
Separate PR for the validation.go fixes: https://github.com/kubernetes/kubernetes/pull/27094

**Comment by thockin** [comment]
I think we are open to getting tests in before 1.3 as long as the test is stable.  

I don't know off hand what is failing here, though - could this be related to needing to enable the beta API in --runtime-config?

**Comment by caseydavenport** [comment]
@thockin @girishkalele OK, so I've removed the failing test for now so that we can get the bits that are working in while I deal with the failing test.  The rest are all passing reliably.

**Comment by matchstick** [comment]
I signed it!

**Comment by matchstick** [comment]
I wanted to verify the tests work before LGTM. After conferring with the wise @mikedanese I performed the following steps:
git fetch https://github.com/caseydavenport/kubernetes.git cd-networkpolicy-uts
git co FETCH_HEAD
go test -v -count 12 ./pkg/apis/extensions/validation ./pkg/registry/networkpolicy/etcd ./pkg/registry/networkpolicy | tee /tmp/bobo
grep FAIL /tmp/bobo

Looks good.

**Comment by caseydavenport** [comment]
@matchstick Just noticed this wasn't added to the v1.3 milestone - any chance of getting a milestone added to this?

**Comment by matchstick** [comment]
Thanks @mikedanese. yes this makes sense to add. @thockin  wanted these tests in since the functionality was added in v1.3.

**Comment by fejta** [comment]
@k8s-bot node e2e test this issue #24479

**Pull request merged**